File: bank/bank.go
package bank

import (
	"errors"

	"github.com/replicasystem/structs"
)

type transaction struct {
	tmap map[string]string
}

func inittransaction() *transaction {
	var t transaction
	t.tmap = make(map[string]string)
	return &t
}

func (t *transaction) recordtransaction(reqid, typet string) {
	t.tmap[reqid] = typet
}

func (t *transaction) checktransaction(reqid, typet string) string {
	if name, ok := t.tmap[reqid]; ok {
		if name == typet {
			return "processed"
		} else {
			return "inconsistent"
		}
	}
	return "new"
}

type account struct {
	accountid string
	balance   int
}

func (a *account) getbalance() int {
	return a.balance
}

func (a *account) deposit(amount int) {
	a.balance = a.balance + amount
}

func (a *account) withdraw(amount int) error {
	temp := a.balance - amount
	if temp < 0 {
		err := errors.New("no funds")
		return err
	}
	a.balance = temp
	return nil
}

type Bank struct {
	amap     map[string]account
	bankname string
	bankid   string
	t        *transaction
}

func Initbank(name, Id string) *Bank {
	var b = Bank{
		amap:     make(map[string]account),
		bankname: name,
		bankid:   Id,
		t:        inittransaction(),
	}
	return &b
}

func (b *Bank) CheckId(accountId string) {
	if _, ok := b.amap[accountId]; !ok {
		var newaccnt = account{
			accountid: accountId,
			balance:   0,
		}
		b.amap["accountId"] = newaccnt
	}
}

func (b *Bank) Deposit(req *structs.Request) *structs.Reply {
	b.CheckId(req.Account)
	a := b.amap[req.Account]
	resp := b.t.checktransaction(req.Requestid, "deposit")
	if resp == "new" {
		resp = "processed"
		a.deposit(req.Balance)
		b.t.recordtransaction(req.Requestid, "deposit")
	}
	return structs.Makereply(req.Requestid, req.Account, resp, "deposit", a.getbalance())
}

func (b *Bank) Withdraw(req *structs.Request) *structs.Reply {
	b.CheckId(req.Account)
	a := b.amap[req.Account]
	resp := b.t.checktransaction(req.Requestid, "withdraw")
	if resp == "new" {
		resp = "processed"
		b.t.recordtransaction(req.Requestid, "withdraw")
		if err := a.withdraw(req.Balance); err != nil {
			return structs.Makereply(req.Requestid, req.Account, "insufficientfunds", "withdraw", a.getbalance())
		}
	}
	return structs.Makereply(req.Requestid, req.Account, resp, "withdraw", a.getbalance())
}

func (b *Bank) Set(rep *structs.Reply) {
	b.CheckId(rep.ReqID)
	a := b.amap[rep.AccountNum]
	a.balance = rep.Balance
	b.t.recordtransaction(rep.ReqID, rep.Transaction)
}

func (b *Bank) GetBalance(req *structs.Request) *structs.Reply {
	b.CheckId(req.Account)
	a := b.amap[req.Account]
	b.t.recordtransaction(req.Requestid, "getbalance")
	return structs.Makereply(req.Requestid, req.Account, "processed", "getbalance", a.getbalance())
}

File: python/bank.py
from structs import Request

class Bank():
  def __init__(self,name,id):
    self.name = name
    self.bankid=id
    self.amap={}
    self.tmap={}

  def recordtransaction(self,reqid,typet):
    self.tmap[reqid]=typet

  def checktransaction(self,reqid,typet):
    if reqid in self.tmap:
      if self.tmap[reqid]==typet:
        return "processed"
      else :
        return "inconsistent"
    return "new"

  def checkaccountid(self,account):
    if account not in self.amap:
      self.amap[account]=0


  def getbalance(self,request):
    self.checkaccountid(request.accountid)
    return Request(request.requestid,request.accountid,self.amap[request.accountid],"getbalance","processed")


  def deposit(self,request):
    self.checkaccountid(request.accountid)
    resp=self.checktransaction(request.requestid,"deposit")
    if resp == "new" :
      resp = "processed"
      bal=self.amap[request.accountid]
      bal =bal + request.balance
      self.amap[request.accountid]=bal
      self.tmap[request.requestid]="deposit"
    return Request(request.requestid,request.accountid,self.amap[request.accountid],"deposit",resp)

  def withdraw(self,request):
    self.checkaccountid(request.accountid)
    resp=self.checktransaction(request.requestid,"withdraw")
    if resp == "new":
      resp="processed"
      bal=self.amap[request.accountid]
      if bal-amount < 0 :
        return Request(request.requestid,request.accountid,self.amap[request.accountid],"withdraw","insufficientfunds")
      bal =bal - amount
      self.amap[request.accountid]=bal
      self.tmap[request.requestid]="withdraw"
    return Request(request.requestid,request.accountid,self.amap[request.accountid],"withdraw",resp)

  def set(self,request):
    self.checkaccountid(request.accountid)
    self.amap[request.accountid]=request.balance
    self.tmap[request.requestid]=request.transaction
